### PR TITLE
Autogenerate/autorun Simulator Tests

### PR DESCRIPTION
**Pull request by Champers5000**
# Quality Assurance Checklist 
To make reviews more efficient, please make sure the software feature meets the following standards and check everything off that meets the quality check. Once everything has been checked, the assigned reviewers will begin the review process. _Edit this description to check off the list._

There are exceptions with all guidelines. As long as your decisions are justified, then you are good! Contact the reviewers or the leads about any exceptions. 

## Requirements 
- [x] Followed Coding Style Guide 
- [x] Code Build checks pass 
- [x] No merge conflicts 
- [x] Software feature has test associated with it. 
- [ ] Software feature has documentation for it updated.
- [x] Test provides useful information and uses relevant data to accurately represent BPS 
    - NOTE: If test file already exists, use that one 
- [ ] 2 Members of the team have already reviewed my PR 
    - NOTE: Approver will not look at PR if 2 other members haven't reviewed first 
- [ ] Did you Have Fun?

## Things to Consider 
- [ ] Even if the above are checked, is this the best way of writing my code? 
    - It's possible to write code that works, but are there ways to make code more efficient?

### CODE REVIEW

File: BSP/Simulator/Data/critical_overtemp.json [hunk 1]
```diff
@@ -0,0 +1,59 @@
+[
```
Comment (manthanand) [inline BSP/Simulator/Data/critical_overtemp.json:1]: I apologize for the confusion. We will need an overtemperature test for charging and discharging. This means a total of 10 trials. For charging, have a temperature higher than 45000 and a negative current. I will update this in the one note as well.

Comment (Champers5000) [reply BSP/Simulator/Data/critical_overtemp.json:1]: k will do

File: BSP/Simulator/Src/Test_Generator.c [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+}
+
+/**
+ * @brief   Generates random numbers between a range
+ * @param   lower The lower bound, inclusive
+ * @param   upper The upper bound, exclusive
+ * @return  Random number between lower and upper
+ */
+int randomRange(int lower, int upper){
+    return (rand() % (upper - lower + 1)) + lower;
```
Comment (connorl309) [inline BSP/Simulator/Src/Test_Generator.c:161]: Is rand() ever seeded? srand(time(NULL))

**Review by connorl309 — COMMENTED** [review]
Overall looks fine but the test files seem a bit... similar to me. I don't think rand() is ever being seeded in the test generator.

File: BSP/Inc/Test_Generator.h [hunk 3]
```diff
@@ -0,0 +1,33 @@
+#ifndef TESTGENERATOR_H
+#define TESTGENERATOR_H
+#define jsonLength 10 //length of the generated jsonfiles
```
Comment (ray4477) [inline BSP/Inc/Test_Generator.h:3]: All caps this

File: BSP/Simulator/Src/Test_Generator.c [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+            break;
+        case '5':
+            index = randomRange(1,jsonLength);
+            state[index].current = MAX_CURRENT_LIMIT + 1;
+            writeOut("Test-OverCurrent");
+            break;
+        case '6':
+            index = randomRange(1,jsonLength);
+            victim = randomRange(0, NUM_TEMPERATURE_SENSORS);
+            state[0].temperatures[victim] = MAX_CHARGE_TEMPERATURE_LIMIT + 1;
```
Comment (ray4477) [inline BSP/Simulator/Src/Test_Generator.c:91]: Should be state[index]

File: BSP/Simulator/Src/Test_Generator.c [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+            writeOut("Test-ChargingOverTemperature");
+            break;
+        case '7':
+            index = randomRange(1,jsonLength - 1);
+            victim = randomRange(0, NUM_TEMPERATURE_SENSORS);
+            state[0].temperatures[victim] = MAX_DISCHARGE_TEMPERATURE_LIMIT + 1;
```
Comment (ray4477) [inline BSP/Simulator/Src/Test_Generator.c:97]: Same here state[index]

File: BSP/Simulator/Src/Test_Generator.c [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+            break;
+        default:
+            break;
+    }
+}
+
+void initializeVariables(bool charging){
+    int flipcurrent = charging ? -1 : 1; //we need to set the curreent to a negative value if the car is charging
+    for(int i=0; i<jsonLength; i++) {    
+        state[i].time = 10+2*i; //first state time is 10, increase by 2 for each state
```
Comment (ray4477) [inline BSP/Simulator/Src/Test_Generator.c:134]: Why? This makes every state after last increasingly more long

**Review by ray4477 — CHANGES_REQUESTED** [review]
Looks good for making test cases, fix bugs though

**Review by ClarkPoon — COMMENTED** [review]
Why did you add the .gch file? Generally, we try to avoid adding object files to version control if we can avoid it since we can build them from the source files

File: BSP/Simulator/Src/Test_Generator.c [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+ *          B-No Error Test
+ *          C-All Edge Cases Error Test
+ * @return  None
+ */
+void generateTest(char selection){
+    int index, victim; //temperary indexing variables
+    //initialization
+    switch(selection){
+        case '1':
+        case '6':
```
[the remark this answers is not in the document]
Comment (Champers5000) [reply BSP/Simulator/Src/Test_Generator.c:133]: The user must pass in these chars to specify which tests, at which point they would just be easier to process as simply chars instead of adding another layer of abstraction. Less typing too

File: BSP/Simulator/Src/Test_Generator.c [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+            initializeVariables(true);
+            break;
+        case '2':
+        case '7':
+            initializeVariables(false);
+            break;
+        default:
+            //default case we randomly choose whether charging or discharging
+            rand()%2==0 ? initializeVariables(true) : initializeVariables(false);
+            break;
```
[the remark this answers is not in the document]
Comment (Champers5000) [reply BSP/Simulator/Src/Test_Generator.c:143]: If we can always check what the simulator fails on because the generated output is written to a file. Some tests like overvoltage should produce the same results regardless of whether we are charging, so we can just randomly pick either charging/discharging to test on

File: BSP/Inc/Test_Generator.h [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+ * @brief   Generates random numbers between a range
+ * @param   lower The lower bound, inclusive
+ * @param   upper The upper bound, exclusive
+ * @return  Random number between lower and upper
+ */
+int randomRange(int lower, int upper);
+
+bool writeOut(char* filename);
+
+void runTests(int input);
```
[the remark this answers is not in the document]
Comment (Champers5000) [reply BSP/Inc/Test_Generator.h]: Oh, I thought it was just common practice to put all method headers into the header file

File: BSP/Simulator/Src/Test_Generator.c [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+
+        for(int j=0; j<NUM_BATTERY_MODULES; j++) {
+            state[i].voltages[j] = randomRange(MIN_VOLTAGE_LIMIT, MAX_VOLTAGE_LIMIT);
+        }
+
+        for(int j=0; j<NUM_TEMPERATURE_SENSORS; j++){
+            state[i].temperatures[j] = randomRange(0, MAX_CHARGE_TEMPERATURE_LIMIT);
+        }
+
+        state[i].current = flipcurrent * randomRange(0, MAX_CURRENT_LIMIT);
```
[the remark this answers is not in the document]
Comment (Champers5000) [reply BSP/Simulator/Src/Test_Generator.c]: We could also just set all batteries to the same voltage , current, and temperature, but that's not as fun. Here I'm generating random numbers within operating limits to fill those values

File: BSP/Simulator/Src/Test_Generator.c [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+ * @brief   Generates random numbers between a range
+ * @param   lower The lower bound, inclusive
+ * @param   upper The upper bound, exclusive
+ * @return  Random number between lower and upper
+ */
+int randomRange(int lower, int upper){
+    return (rand() % (upper - lower + 1)) + lower;
+}
+
+bool writeOut(char* filename){
```
[the remark this answers is not in the document]
Comment (Champers5000) [reply BSP/Simulator/Src/Test_Generator.c:252]: Too late for that, I already wrote it ¯\_(ツ)_/¯

File: BSP/Simulator/Makefile [hunk 12]
```diff
@@ -12,7 +12,7 @@ CC = gcc
 ######################################
 # Source
 ######################################
-SRC = $(wildcard Src/*.c)	# BSP code
+SRC = $(filter-out Src/Test_Generator.c, $(wildcard Src/*.c))	# BSP code
```
[the remark this answers is not in the document]
Comment (Champers5000) [reply BSP/Simulator/Makefile:15]: That was just a happy little accident that it was named Test_Generator. There's a main method in there, so it has to be compiled separate from the simulator main method, hence changes in the makefile

File: BSP/Simulator/Src/Test_Generator.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (manthanand) [reply BSP/Simulator/Src/Test_Generator.c:133]: You could just `#define CRITICAL_OVERTEMP_TEST '1'` or something.

**Comment by manthanand** [comment]
I don't think this PR is perfect in that it could be more readable (using CJSON library which has documentation and examples, using enums to make things easier to understand), but the functionality works. If you would like to save these changes for some other time, make issue tickets on the GitHub, and I will merge the PR.

File: BSP/Simulator/Makefile (hunk 12 above)
[the remark this answers is not in the document]
Comment (manthanand) [reply BSP/Simulator/Makefile:15]: Could you rename to Simalator_GenerateTests? Makes more sense and won't be confusing.

File: .github/workflows/simple-tests.yml [hunk 13]
```diff
@@ -25,13 +25,14 @@ jobs:
...
-        sudo ./Validation/test.sh critical_overcurrent
-        sudo ./Validation/test.sh critical_overtemp_discharging
-        sudo ./Validation/test.sh critical_overvolt
-        sudo ./Validation/test.sh critical_undervolt
-        sudo ./Validation/test.sh overcurrent
-        sudo ./Validation/test.sh overtemp_charging
-        sudo ./Validation/test.sh overtemp_discharging
-        sudo ./Validation/test.sh overvolt
-        sudo ./Validation/test.sh undervolt
+        sudo bash Validation/test.sh BSP/Simulator/Data/ManualTests/critical_overcurrent.json
```
Comment (manthanand) [inline .github/workflows/simple-tests.yml:28]: Couldn't the file path be moved into the test.sh file? would make things a bit nicer to read.

File: BSP/Inc/Automatic_Tester.h [hunk 14]
```diff
@@ -0,0 +1,73 @@
+#ifndef TESTGENERATOR_H
```
Comment (manthanand) [inline BSP/Inc/Automatic_Tester.h:1]: Need copyright header and description of library over here.

File: BSP/Simulator/Src/Automatic_Tester.c [hunk 15]
```diff
@@ -0,0 +1,376 @@
+#include "Automatic_Tester.h"
```
Comment (manthanand) [inline BSP/Simulator/Src/Automatic_Tester.c:1]: copyright headers and stuff

**Comment by manthanand** [comment]
Also, did you run the test that failed the overcurrent random generated test?

File: BSP/Inc/Automatic_Tester.h [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+ * @note If no command line arguments are supplied, you will have to type them after the program starts running
+ * 
+*/
+int main(int argc, char **argv);
+
+/**
+ * @brief   Generates the test json, outputs it to file test.json
+ * @param   selection Selects which test to generate
+ *          0-Critical Overcurrent
+ *          1-Critical Overtemp while charging
```
Comment (manthanand) [inline BSP/Inc/Automatic_Tester.h:31]: We arent charging in critical state so 1 should not be a valid input.

File: BSP/Inc/Automatic_Tester.h [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+ * @param charging Whether or not the car is in a charging state
+*/
+void initializeVariables(bool charging);
+
+/**
+ * @brief   Writes out a json file using all the variables set earlier
+ * @param   filename The name of the json to write out to, filepath and .json extension will be added automatically
+ * @return  Whether the creation of json was successful
+*/
+bool writeOut(char* filename);
```
Comment (manthanand) [inline BSP/Inc/Automatic_Tester.h:61]: Make the outputs void instead of bool since it isn't used anywhere.

File: BSP/Inc/Automatic_Tester.h [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+void runTest(char input);
+
+/**
+ * @brief Writes the expected output of the simulator
+ * @param filename The filename that the output file will have
+ * @note The method will automatically which test generate the expected output by the filename
+*/
+bool writeVerification(char* filename);
+
+#endif
```
Comment (manthanand) [inline BSP/Inc/Automatic_Tester.h:76]: Add newline

File: BSP/Simulator/Src/Automatic_Tester.c [hunk 19]
```diff
@@ -0,0 +1,386 @@
...
+        asprintf(&command, " python3 Validation/verify_test.py `find BSP/Simulator/Data/AutomatedTests/ -name \"%c-*\" ! -name \"*-out.json\"`", input); //generate command
+        printf("%s\n", command);
+        if(system(command) != 0){
+           //check if test was successful, if not we need to throw error and exit
+           system("cat BSP/Simulator/Simulator-Out/%c-*");
+           system("echo -e \"The json file used was \n\"; cat BSP/Simulator/Data/AutomatedTests/%c-*");
+           exit(-1); 
+        } 
+    }
+}
```
Comment (manthanand) [inline BSP/Simulator/Src/Automatic_Tester.c:386]: Add newline

File: BSP/STM32F413/Makefile [hunk 20]
```diff
@@ -239,13 +239,13 @@ $(BUILD_DIR):
 .INTERMEDIATE: build_msg
 
 c_compile_msg:
-	@echo -e "${GREEN}Compiling C files...${NC}"
```
Comment (Champers5000) [inline BSP/STM32F413/Makefile]: why are you getting rid of the -e option here? It's needed for escape characters to work

**Review by Champers5000 — COMMENTED** [review]
Looks like it's working

File: BSP/Simulator/Src/Automatic_Tester.c [hunk 21]
```diff
@@ -0,0 +1,403 @@
...
+    FILE *f;
+    f = fopen(filepath, "w");
+    if (f == NULL) {
+        fprintf(stderr, "can't open %s: %s\n", filepath, strerror(errno));
+        return false;
+    }
+    fprintf(f, "[\n");
+    for (int i = 0; i < jsonLength; i++) {
+        fprintf(f, "    {\n");
+        fprintf(f, "        \"time\": %d,\n", states[i].time);
```
Comment (connorl309) [inline BSP/Simulator/Src/Automatic_Tester.c:278]: might want to change these to \t\t but i dont think it makes a big difference

File: Validation/automatedTest.sh [hunk 22]
```diff
@@ -0,0 +1,5 @@
+ulimit -r unlimited
```
Comment (connorl309) [inline Validation/automatedTest.sh:1]: needs to be sudo i think

**Review by connorl309 — COMMENTED** [review]
a couple changes nothing major, i am trusting that the tester.c file works because that has some scary lower level calls in it that make me a bit worried. the json formatting looks good

File: Validation/automatedTest.sh (hunk 22 above)
Comment (manthanand) [reply Validation/automatedTest.sh:1]: It doesn't. The script is run with sudo

**Review by manthanand — APPROVED** [review]

**Review by manthanand — APPROVED** [review]
